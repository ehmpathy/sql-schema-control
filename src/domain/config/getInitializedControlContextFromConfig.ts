/*
  1. get the config (validated and hydrated)
  2. initialize the environment based on the config
    - start the db connection
    - initialize the control environment
  3. return the language+dialect, connection, and definitions
*/
