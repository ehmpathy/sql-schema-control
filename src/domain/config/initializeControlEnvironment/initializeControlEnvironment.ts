/*
  complete any actions required to initialize control environment
    1. connect to db
    2. provision change log table
*/
