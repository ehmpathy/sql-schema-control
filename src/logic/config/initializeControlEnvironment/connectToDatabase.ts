import mysql from 'mysql2/promise';
import { DatabaseConnection, DatabaseLanguage, ControlConfig, ConnectionConfig } from '../../../types';

// create the connection to database
const connectionAdapters = {
  [DatabaseLanguage.MYSQL]: async ({ connectionConfig }: { connectionConfig: ConnectionConfig }): Promise<DatabaseConnection> => {
    const connection = await mysql.createConnection({
      host: connectionConfig.host,
      port: connectionConfig.port,
      database: connectionConfig.schema,
      user: connectionConfig.username,
      password: connectionConfig.password,
    });
    return {
      query: async ({ sql, values }: { sql: string, values?: any[] }) => {
        return connection.query(sql, values);
      },
      end: async () => connection.end(),
    };
  },
};

export const connectToDatabase = async ({ config }: { config: ControlConfig }): Promise<DatabaseConnection> => {
  // 1. get the connection adapter for the method
  const getDbConnection = connectionAdapters[config.language];

  // 2. attempt to connect
  const connection = await getDbConnection({ connectionConfig: config.connection });

  // 3. run a test query to check connection
  await connection.query({ sql: 'SHOW TABLES' }); // TODO: make this not mysql specific

  // 4. return the connection
  return connection;
};
