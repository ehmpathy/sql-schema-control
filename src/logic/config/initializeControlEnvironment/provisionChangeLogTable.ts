import { DatabaseConnection } from '../../../types';
import { readFileAsync } from '../_utils/readFileAsync';

export const provisionChangeLogTable = async ({ connection }: { connection: DatabaseConnection }) => {
  // create the table if not exists (sql consideres its existance for us)
  const createTableSql = await readFileAsync({
    filePath: `${__dirname}/../../../../schema/tables/schema_control_change_log.mysql.sql`,
  }); // TODO: make independent of language (mysql currently)
  await connection.query({ sql: createTableSql });
};
