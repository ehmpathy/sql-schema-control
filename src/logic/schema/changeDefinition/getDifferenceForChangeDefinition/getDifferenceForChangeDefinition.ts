import { DatabaseConnection, ChangeDefinition } from '../../../../types';
import { getSqlDifference } from '../../utils/getSqlDifference';
import { getAppliedChangeDefinitionFromDatabase } from '../getAppliedChangeDefinitionFromDatabase/getAppliedChangeDefinitionFromDatabase';

export const getDifferenceForChangeDefinition = async ({
  connection,
  change,
}: {
  connection: DatabaseConnection;
  change: ChangeDefinition;
}) => {
  // 1. get state of change in db
  const lastAppliedChange = await getAppliedChangeDefinitionFromDatabase({
    connection,
    changeId: change.id,
    changePath: change.path,
  });

  // 2. cast into string
  const sqlDifference = getSqlDifference({ oldSql: lastAppliedChange.sql, newSql: change.sql });

  // 3. return the diff string
  return sqlDifference;
};
