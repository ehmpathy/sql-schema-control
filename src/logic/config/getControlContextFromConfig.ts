import { ControlContext, ResourceDefinition } from '../../types';
import { getUncontrolledResources } from '../schema/resourceDefinition/getUncontrolledResources';
import { getConfig } from './getConfig';
import { initializeControlEnvironment } from './initializeControlEnvironment';

/*
  1. get the config (validated and hydrated)
  2. initialize the environment based on the config
    - start the db connection
    - initialize the control environment
  3. return the control context
*/
export const getControlContextFromConfig = async ({
  configPath,
  strict = false,
}: {
  configPath: string;
  strict?: boolean;
}) => {
  // 1. get the config
  const config = await getConfig({ configPath });

  // 2. initialize the environment based on the config
  const { connection } = await initializeControlEnvironment({ config });

  // 3. define the definitions such that we can append to them if we want (e.g., if want to add uncontrolled)
  const definitions = [...config.definitions];

  // 4. get uncontrolled resources, if strict
  if (config.strict || strict) {
    const controlledResources = definitions.filter(
      (def) => def.constructor === ResourceDefinition,
    ) as ResourceDefinition[];
    const uncontrolledResources = await getUncontrolledResources({ connection, controlledResources });
    definitions.push(...uncontrolledResources); // append the uncontrolled resources to the definitions
  }

  // 5. return the control context
  return new ControlContext({
    language: config.language,
    dialect: config.dialect,
    connection, // note: the connection is a live dbConnection
    definitions, // note: uncontrolled resources will already have statuses
  });
};
