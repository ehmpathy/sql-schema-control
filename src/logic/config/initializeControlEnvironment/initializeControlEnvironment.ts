import { ControlConfig } from '../../../domain';
import { connectToDatabase } from './connectToDatabase';
import { provisionChangeLogTable } from './provisionChangeLogTable';

/*
  complete any actions required to initialize control environment
    1. connect to db
    2. provision change log table
*/
export const initializeControlEnvironment = async ({
  config,
}: {
  config: ControlConfig;
}) => {
  // 1. connect to db
  const connection = await connectToDatabase({ config });

  // 2. provision the change log table
  await provisionChangeLogTable({ connection });

  // 3. return the connection
  return { connection };
};
