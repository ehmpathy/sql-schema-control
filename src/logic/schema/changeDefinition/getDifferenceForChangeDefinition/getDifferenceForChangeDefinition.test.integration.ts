import sha256 from 'simple-sha256';
import uuid from 'uuid/v4';
import {
  DatabaseLanguage,
  DatabaseConnection,
  ControlConfig,
  ChangeDefinition,
  ChangeDefinitionStatus,
} from '../../../../types';
import { initializeControlEnvironment } from '../../../config/initializeControlEnvironment';
import { getDifferenceForChangeDefinition } from './getDifferenceForChangeDefinition';
import { promiseConfig } from '../../../../__test_assets__/connection.config';

describe('getDifferenceForChangeDefinition', () => {
  let connection: DatabaseConnection;
  beforeAll(async () => {
    const config = new ControlConfig({
      language: DatabaseLanguage.MYSQL,
      dialect: '5.7',
      connection: (await promiseConfig()).mysql,
      definitions: [],
      strict: true,
    });
    ({ connection } = await initializeControlEnvironment({ config })); // ensure db is provisioned and get connection
  });
  afterAll(async () => {
    await connection.end();
  });
  it('should find that a change that is up to date has no difference', async () => {
    // define the change
    const sql = `
    CREATE TABLE IF NOT EXISTS some_table (
      id              INT(11) PRIMARY KEY AUTO_INCREMENT,
      created_at      DATETIME DEFAULT CURRENT_TIMESTAMP,
    ) ENGINE = InnoDB;
    `;
    const definition = new ChangeDefinition({
      id: uuid(),
      path: '__PATH__',
      sql,
      hash: sha256.sync(sql),
    });

    // apply it
    await connection.query({
      sql: `
      INSERT INTO schema_control_change_log
        (change_id, change_hash, change_content)
      VALUES
        ('${definition.id}', '${definition.hash}', '${definition.sql}');
    `,
    });

    // show that there is no diff
    const result = await getDifferenceForChangeDefinition({ connection, change: definition });
    expect(result).toEqual(null);
  });
  it('should find that an out of date change has a colored diff of changed lines', async () => {
    // record that the definition was already applied
    const definition = new ChangeDefinition({
      id: uuid(),
      path: '__PATH__',
      sql: `
CREATE TABLE IF NOT EXISTS some_table (
  id              INT(11) PRIMARY KEY AUTO_INCREMENT,
  created_at      DATETIME DEFAULT CURRENT_TIMESTAMP,
) ENGINE = InnoDB;
      `,
      hash: sha256.sync(''),
    });
    await connection.query({
      sql: `
      INSERT INTO schema_control_change_log
        (change_id, change_hash, change_content)
      VALUES
        ('${definition.id}', '${definition.hash}', '${definition.sql}');
    `,
    });

    // update the original definition
    const updatedDefinition = new ChangeDefinition({
      ...definition,
      sql: `
CREATE TABLE IF NOT EXISTS some_table (
id              BIGINT PRIMARY KEY AUTO_INCREMENT,
created_at      DATETIME(3) DEFAULT CURRENT_TIMESTAMP
) ENGINE = InnoDB;
      `,
      hash: sha256.sync('2'),
      status: ChangeDefinitionStatus.OUT_OF_DATE,
    });

    // get the diff
    const result = await getDifferenceForChangeDefinition({ connection, change: updatedDefinition });
    expect(typeof result).toEqual('string');
    expect(result).toMatchSnapshot(); // result is purely visual, so log an example of it
  });
});
