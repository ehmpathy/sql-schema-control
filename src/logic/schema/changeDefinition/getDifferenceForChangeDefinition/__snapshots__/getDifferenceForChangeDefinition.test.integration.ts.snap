// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDifferenceForChangeDefinition should find that an out of date change has a colored diff of changed lines 1`] = `
"[90m[39m
[90mCREATE TABLE IF NOT EXISTS some_table ([39m
[90m[39m[31mid              INT(11) PRIMARY KEY AUTO_INCREMENT,[39m
[31mcreated_at      DATETIME DEFAULT CURRENT_TIMESTAMP,[39m
[31m[39m[32mid              BIGINT PRIMARY KEY AUTO_INCREMENT,[39m
[32mcreated_at      DATETIME(3) DEFAULT CURRENT_TIMESTAMP[39m
[32m[39m[90m) ENGINE = InnoDB;[39m
[90m[39m"
`;
